hwdb: fix trie search

cr8t opened the pull request:
Fixes broken HWDB trie search algorithm.

Trie node and entry parsing now uses sizes parsed from the HWDB header.

The inner index in `trie_search` is updated properly following the `eudev` implementation.

Further refactoring is needed to bring the implementation closer to idiomatic Rust.

Includes fixes from John Whittington <git@jbrengineering.co.uk> (many thanks :)

Resolves: https://github.com/cr8t/udev/issues/22

tuna-f1sh commented:
Yes it works as expected for the [libudev port](https://github.com/systemd/systemd/blob/main/src/libudev/libudev-hwdb.c#L105) - thanks. You're right I also did some digging and documentation is minimal. I looked at [systemd sd-hwdb](https://github.com/systemd/systemd/blob/main/src/libsystemd/sd-hwdb/sd-hwdb.c) as a reference. If you use 

```
❯ systemd-hwdb query usb:v1D6Bp0001
ID_VENDOR_FROM_DATABASE=Linux Foundation
ID_AUTOSUSPEND=1
ID_MODEL_FROM_DATABASE=1.1 root hub
```
One gets all the key/values too though.

It's not that idiomatic Rust though as you say. I have a couple of proposals:

* Add query helpers like [udev-rs](https://github.com/Smithay/udev-rs/blob/master/src/hwdb.rs#L48) since the current `udev_hwdb_get_properties_list_entry` then calling `property_list()` is non-obvious. https://github.com/tuna-f1sh/udev/commit/e3c5d78bc29d15abe8cf570e77c6e7a99324033e
* Change return of `udev_hwdb_get_properties_list_entry` to be Iter (reasoning in commit message). I think the change to the test shows why this is nicer but note it would be a breaking change: https://github.com/tuna-f1sh/udev/commit/d27ba56a4c229aa4b89c016f392f0966012a6969

cr8t commented:
> Add query helpers like [udev-rs](https://github.com/Smithay/udev-rs/blob/master/src/hwdb.rs#L48) since the current udev_hwdb_get_properties_list_entry then calling property_list() is non-obvious. https://github.com/tuna-f1sh/udev/commit/e3c5d78bc29d15abe8cf570e77c6e7a99324033e

I like this solution. I've moved the helpers into `UdevHwdb` methods, and changed one signature to return a borrowed `&str` value.

Thank you for all your help on this issue.

tuna-f1sh commented:
Great thanks. Would be good to get a release with fix when you get chance too.
